difflint/lib.py

```diff
@@ -196,6 +216,58 @@ def detect_new_diff_lint_errors(log_output):
             return True
     return False
 
+def save_merge_state():
+    """Saves the merge state in the event that we're resolving a
+    merge conflict. Stashing will mangle the merge state unless
+    we have saved it ahead of time. The reverse operation to this
+    function is restore_merge_state().
+
+    Special thanks to the overcommit project for solving this
+    problem in a similar context. (https://github.com/brigade/overcommit)
+
+    Inputs: None
+
+    Output: If we were in a merge conflict state, this will return a
+            tuple containing first the merge message, and second the
+            merge commit hash.
+            If we were not in a merge conflict state, this will return a
+            tuple containing first an empty string, and second another
+            empty string.
+    """
+    merge_head_commit_hash = ''
+    try:
+        # This will fail if MERGE_HEAD doesn't exist.
+        merge_head_commit_hash = check_output(['git', 'rev-parse', 'MERGE_HEAD'],
+                                              stderr=DEVNULL).decode()
+    except CalledProcessError:
+        return ('', '')
+
+    merge_msg = ''
+
+    merge_msg_path = repo_root() / '.git' / 'MERGE_MSG'
+    if merge_msg_path.is_file():
+        merge_msg = merge_msg_path.read_text()
+
+    return (merge_msg, merge_head_commit_hash)
+
+def restore_merge_state(merge_msg, merge_head_commit_hash):
+    """Restores the merge state saved by a previous call to
+    save_merge_state(). This function should only be called
+    if save_merge_state() returned a non-empty string.
+
+    Inputs:
+        merge_msg: The saved message from a merge conflict
+
+        merge_head_commit_hash: The hash belonging to the
+            MERGE_HEAD reference.
+
+    Output: None
+    """
+    dot_git = repo_root() / '.git'
```

Comment:
I'm pretty sure that the slowness is due to swapping files in and out of the repo when you stash and unstash. Note that git probably does the exact same thing that we have implemented in `repo_root()` every time you run a git command.